### PR TITLE
Stacked timer reporting

### PR DESCRIPTION
**Pull request by rppawlo**
Fixed two issues. Each issue is addressed in a separate commit. Might be easier to review by commit.

1. Addressed #4119 - stacked timer was seg faulting when using with kokkos space_time_stack profiling tool. This has been fixed. Users can control the start of the top level timer by calling startBaseTimer() or let the StackedTimer ctor automatically start it. Default is for ctor to start it automatically for backwards compatibility. The space_time_stack will seg fault if the last timer is not stopped. This was accomplished by calling stop() on the stacked timer. This function has been renamed to be consistent with the startBaseTimer() and to better describe what it really does.

1. Fixed an issue with the parallel report() function. It was not flattening the 0 level timer correctly at the end of the recursion, so the level 0 and level 1 timers were printed on the same level.

### CODE REVIEW

File: packages/teuchos/comm/test/StackedTimer/stacked_timer.cpp [hunk 1]
```diff
@@ -335,4 +344,21 @@ TEUCHOS_UNIT_TEST(StackedTimer, OverlappingTimersViaRCP)
...
+  // Note that the dtor for GlobalMPISession will call
+  // Kokkos::finalize_all().
+  Teuchos::GlobalMPISession mpiSession(&argc, &argv);
+  Kokkos::initialize(argc,argv);
+  {
+    Teuchos::FancyOStream out(Teuchos::rcpFromRef(std::cout));
+    out.setOutputToRootOnly(0);
+  }
+  Teuchos::UnitTestRepository::setGloballyReduceTestResult(true);
+  return Teuchos::UnitTestRepository::runUnitTestsFromMain(argc, argv);
```
[the remark this answers is not in the document]
Comment (rppawlo) [reply packages/teuchos/comm/test/StackedTimer/stacked_timer.cpp]: The main() creates a Teuchos::GlobalMPISession. This has a Kokkos::finalize_all() in the destructor, so finalize is being called. The ScopeGuard is more explicit. I could add that instead, but then finalize will be called twice. Not sure if that causes issues. I can test. What would you prefer?

File: packages/teuchos/comm/src/Teuchos_TimeMonitor.cpp [hunk 2]
```diff
@@ -260,7 +260,7 @@ namespace Teuchos {
       counter().start(reset);
 #ifdef HAVE_TEUCHOS_ADD_TIME_MONITOR_TO_STACKED_TIMER
       if (nonnull(stackedTimer_))
-        stackedTimer_->start(counter().name());
+        stackedTimer_->start(counter().name(),false);
```
Comment (bathmatt) [inline packages/teuchos/comm/src/Teuchos_TimeMonitor.cpp]: So, if you use teuchos you don't get the kokkos injection but you do if you use stacked timer directly?  Should they be the same?

**Comment by bathmatt** [comment]
I'm fine w/ this, see my one comment..

File: packages/teuchos/comm/src/Teuchos_TimeMonitor.cpp (hunk 2 above)
Comment (rppawlo) [reply packages/teuchos/comm/src/Teuchos_TimeMonitor.cpp]: This prevents double injection into kokkos. The TimeMonitor timers inject into the kokkos profiler inside the Time object. But if TimeMonitor injects data into StackedTimer as well, we were doubly injecting into kokkos.

File: packages/teuchos/comm/test/StackedTimer/stacked_timer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhoemmen) [reply packages/teuchos/comm/test/StackedTimer/stacked_timer.cpp]: @rppawlo Didn't realize that `GlobalMPISession` calls `finalize_all`.  Does it also call `Kokkos::initialize`?  If not, is there some reason why it wouldn't want to?

[the remark this answers is not in the document]
Comment (rppawlo) [reply packages/teuchos/comm/test/StackedTimer/stacked_timer.cpp]: It does not call initialize. I think the reason is mostly historical. This was added in the early days of kokkos (Jan 2015 according to git blame). It was done to fix shutdown warnings when Tpetra was originally converted to kokkos data structures. I think back then, the initialize function was in flux.

[the remark this answers is not in the document]
Comment (rppawlo) [reply packages/teuchos/comm/test/StackedTimer/stacked_timer.cpp]: @mhoemmen - your change request is blocking merge. Can I mark this resolved?

**Comment by mhoemmen** [comment]
@rppawlo You might want to add a Kokkos::finalize at the end, conditional on `Kokkos::is_initialized()`.  If we ever fix #4142, then your test won't break.

**Comment by rppawlo** [comment]
> You might want to add a Kokkos::finalize at the end

@mhoemmen - I've added this but now need to be approved again due to change. Could you do this?

**Pull request merged**